lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -684,6 +794,77 @@ def _set_unit_ip(self, event):
                 self._charm.model.get_binding(relation).network.bind_address
             )
 
+    def _label_alert_topology(self, rule):
+        """Insert juju topology labels into an alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing Prometheus alert rule with Juju
+            topology labels.
+        """
+        metadata = self._scrape_metadata
+        labels = rule.get("labels", {})
+        labels["juju_model"] = metadata["model"]
```

Comment:
We should not override these labels if they exist. We should have the equivalent of the `honor_labels = true` semantics.

Reply:
I foresee some very unlikely corner cases involving remapping models for Bootstack where this may matter. Overall, I think it is good practice for us to be mindful of generally not plainly overriding metadata.

Reply:
If we do this there are no guarantees on the sanity of these labels. I do not see why a consumer charm would like to set Juju topology labels explicitly. From what you have suggested uptil now I understood that you wanted these labels to be added automatically.

Reply:
Could we consider changing this behaviour when we actually hit those corner cases and can not find any way out ? For the moment could we advertise the behaviour as "Please do not set Juju Topology labels since these will be set for you automatically".

Reply:
> I do not see why a consumer charm would like to set Juju topology labels explicitly.

How about THIS charm: https://warthogs.atlassian.net/browse/OP-234?atlOrigin=eyJpIjoiODIzN2VmNTg3YWVjNDc3ODgzODEzN2M2NjIyNjkxN2EiLCJwIjoiaiJ9 :D

Reply:
I am also thinking that [Scrape Configuration Charm](https://warthogs.atlassian.net/browse/OP-223?atlOrigin=eyJpIjoiOTU5ODIyZTAyZTI3NDU4Mjg0NGJmNDQ4YjNiZjY3YmEiLCJwIjoiaiJ9) will also need to override the topology labels so that the values of the actually-scraped charm will be used instead.

Reply:
You could also do a conditional assignment as a simple guard against this?

```python
labels["juju_model"] = metadata["model"] if not labels["juju_model"] else labels["juju_model"]
```

Probably makes the code a little more coarse to read, but does get around the issue.

That said, does Balbir have a point here, do we in face want people to be able to override these fields at all?

Reply:
Both Scrape Target and Scrape Configuration could accept `scrape_metadata` from the scrape targets they are related too. In fact `scrape_metadata` which contains the Juju topology is part of the `prometheus_scrape` interface. I have created a [https://github.com/canonical/prometheus-operator/issues/83] to think this through and if necessary refactor the Juju topology labeling workflow in a future pull request.

Reply:
Ah, right. Then I retire my objection :-)